File: runme/mcgen/local/datan/datsrc/mincmb.c
#include "datsrc.h"

void LIBFUNC mincmb_(double P_T *a, double P_T *b, double const P_T *epsiln,
             double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), double P_T *xmin, double P_T *fmin,
             integer P_T *nstep, double const P_T *x0, double const P_T *xdir,
             integer const P_T *n)
{
    /* System generated locals */
    double d__1;

    /* Local variables */
    integer istep;
    double d, e, p, q, r, u, v, w, x, t2, fu, fv, fw, fx, xm, eps, tol;
    logical parstp;

/* if some input values are not given use defaults */
    /* Parameter adjustments */
    --xdir;
    --x0;

    /* Function Body */
    if (*nstep < 1) {
        *nstep = 1000;
    }
    eps = *epsiln;
    if (eps <= 0.) {
        eps = 1e-12;
    }
/* initialize X at golden section position between A and B */
    x = *a + (*b - *a) * .381966;
/* initialize E,V,W,FX,FV,FW */
    e = 0.;
    v = x;
    w = x;
    fx = minfnd_(&x, &x0[1], &xdir[1], n, userfn);
    fv = fx;
    fw = fx;
    istep = 1;
/* start iteration */
L10:
    if (istep >= *nstep) {
/* prepare exit if step number is too large */
        *nstep = -1;
        *xmin = x;
        *fmin = fx;
        goto L20;
    }
    xm = (*a + *b) * .5;
    tol = eps * abs(x) + 1e-15;
    t2 = tol * 2.;
    if ((d__1 = x - xm, abs(d__1)) > t2 - (*b - *a) * .5) {
        p = 0.;
        q = 0.;
        r = 0.;
        parstp = FALSE_;
        if (abs(e) > tol) {
/* fit parabola */
            r = (x - w) * (fx - fv);
            q = (x - v) * (fx - fw);
            p = (x - v) * q - (x - w) * r;
            q = (q - r) * 2.;
            if (q > 0.) {
                p = -p;
            } else {
                q = -q;
            }
            r = e;
            e = d;
            if (abs(p) < (d__1 = q * .5 * r, abs(d__1)) && p > q * (*a - x) &&
                     p < q * (*b - x)) {
/* use result of parabolic fit */
                d = p / q;
                u = x + d;
                if (u - *a < t2 || *b - u < t2) {
/* make sure that U is not too near to A or B */
                    d = -tol;
                    if (x < xm) {
                        d = tol;
                    }
                }
                parstp = TRUE_;
            }
        }
        if (! parstp) {
/* perform golden section step */
            if (x < xm) {
                e = *b - x;
            } else {
                e = *a - x;
            }
            d = e * .381966;
        }
/* determine point U where function is to be computed */
/* making sure that it is not too close to X */
        if (abs(d) >= tol) {
            u = x + d;
        } else if (d > 0.) {
            u = x + tol;
        } else {
            u = x - tol;
        }
        fu = minfnd_(&u, &x0[1], &xdir[1], n, userfn);
/* update A,B,V,W,X */
        if (fu <= fx) {
            if (u < x) {
                *b = x;
            } else {
                *a = x;
            }
            v = w;
            fv = fw;
            w = x;
            fw = fx;
            x = u;
            fx = fu;
        } else {
            if (u < x) {
                *a = u;
            } else {
                *b = u;
            }
            if (fu <= fw || w == x) {
                v = w;
                fv = fw;
                w = u;
                fw = fu;
            } else if (fu <= fv || v == x || v == w) {
                v = u;
                fv = fu;
            }
        }
        ++istep;
        goto L10;
    } else {
        *xmin = x;
        *fmin = fx;
        *nstep = istep;
    }
L20:
    return;
} /* mincmb_ */

File: runme/mcgen/local/datan/datsrc/mincmb.f
      SUBROUTINE MINCMB(A,B,EPSILN,USERFN,XMIN,FMIN,NSTEP,
     +X0,XDIR,N)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DOUBLE PRECISION MINFND
      LOGICAL PARSTP
      PARAMETER(C=0.381966D0,MAXSTP=1000,EPSDEF=1.D-12,TT=1.D-15)
      PARAMETER(ZERO=0.D0, HALF=0.5D0, ONE=1.D0, TWO=2.D0)
      DIMENSION X0(N),XDIR(N)
      EXTERNAL USERFN
C if some input values are not given use defaults
      IF(NSTEP.LT.1) NSTEP=MAXSTP
      EPS=EPSILN
      IF(EPS.LE.ZERO) EPS=EPSDEF
C initialize X at golden section position between A and B
      X=A+C*(B-A)
C initialize E,V,W,FX,FV,FW
      E=ZERO
      V=X
      W=X
      FX=MINFND(X,X0,XDIR,N,USERFN)
      FV=FX
      FW=FX
      ISTEP=1
C start iteration
   10 CONTINUE
      IF(ISTEP.GE.NSTEP) THEN
C prepare exit if step number is too large
        NSTEP=-1
        XMIN=X
        FMIN=FX
        GO TO 20
      END IF
      XM=HALF*(A+B)
      TOL=EPS*ABS(X)+TT
      T2=TWO*TOL
      IF(ABS(X-XM).GT.T2-HALF*(B-A)) THEN
        P=ZERO
        Q=ZERO
        R=ZERO
        PARSTP = .FALSE.
        IF(ABS(E).GT.TOL) THEN
C fit parabola
          R=(X-W)*(FX-FV)
          Q=(X-V)*(FX-FW)
          P=(X-V)*Q-(X-W)*R
          Q=TWO*(Q-R)
          IF(Q.GT.ZERO) THEN
            P=-P
          ELSE
            Q=-Q
          END IF
          R=E
          E=D
          IF(ABS(P).LT.ABS(HALF*Q*R) .AND. P.GT.Q*(A-X) .AND. P.LT.
     +    Q*(B-X)) THEN
C use result of parabolic fit
            D=P/Q
            U=X+D
            IF((U-A).LT.T2 .OR. (B-U).LT.T2) THEN
C make sure that U is not too near to A or B
              D=-TOL
              IF(X.LT.XM) D=TOL
            END IF
            PARSTP = .TRUE.
          END IF
        END IF
        IF(.NOT. PARSTP) THEN
C perform golden section step
          IF(X.LT.XM) THEN
            E=B-X
          ELSE
            E=A-X
          END IF
          D=C*E
        END IF
C determine point U where function is to be computed
C making sure that it is not too close to X
        IF(ABS(D) .GE. TOL) THEN
          U=X+D
        ELSE IF (D.GT.ZERO) THEN
          U=X+TOL
        ELSE
          U=X-TOL
        END IF
        FU=MINFND(U,X0,XDIR,N,USERFN)
C update A,B,V,W,X
        IF(FU.LE.FX) THEN
          IF(U.LT.X) THEN
            B=X
          ELSE
            A=X
          END IF
          V=W
          FV=FW
          W=X
          FW=FX
          X=U
          FX=FU
        ELSE
          IF(U.LT.X) THEN
            A=U
          ELSE
            B=U
          END IF
          IF(FU.LE.FW .OR. W.EQ.X) THEN
            V=W
            FV=FW
            W=U
            FW=FU
          ELSE IF(FU.LE.FV .OR. V.EQ.X .OR. V.EQ.W) THEN
            V=U
            FV=FU
          END IF
        END IF
        ISTEP=ISTEP+1
        GO TO 10
      ELSE
        XMIN=X
        FMIN=FX
        NSTEP=ISTEP
      END IF
   20 CONTINUE
      END
